Improve end to end tests

kmendler opened the pull request:
- Modify test implementation and refactor
- Add test for number of files in output folder
- Add check for annotations table in output folder

Commit:
Modify test implementation and refactor
Add test for number of files in output folder
Add check for annotations table in output folder

LeeBergstrand commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
 
-cd testing/test_constants
+EXIT_CODE=0
+ROOT_PATH=$(pwd)
```
I think that ```PROJECT_ROOT_PATH``` would be a more appropriate variable name. ```ROOT_PATH``` could potentially imply that this path is the path to the filesystem root (i.e. ```/```).

LeeBergstrand commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
-# Get reference hash file checksums
-FIRST_HASH=$(md5sum rpoB_0_msa_0.fa | cut -d ' ' -f 1)
-SECOND_HASH=$(md5sum rpoB_0_refseq_msa_1.fa | cut -d ' ' -f 1)
+## Check that the number of files in --output_dir is correct
```
```Check that the number of files in --output_dir is correct``` --> Mentioning that this the output directory for Metannotate (not one of its individual sub-tools) would be more explicit.

LeeBergstrand commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
-FIRST_HASH=$(md5sum rpoB_0_msa_0.fa | cut -d ' ' -f 1)
-SECOND_HASH=$(md5sum rpoB_0_refseq_msa_1.fa | cut -d ' ' -f 1)
+## Check that the number of files in --output_dir is correct
+# TODO: replace this test with individual tests for each expected file
```
Capital ```R``` on replace.

LeeBergstrand commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
+cd ${ROOT_PATH}/testing/test_constants
+REFERENCE_FILE_HASHES[0]=$(md5sum rpoB_0_msa_0.fa | cut -d ' ' -f 1)
+REFERENCE_FILE_HASHES[1]=$(md5sum rpoB_0_refseq_msa_1.fa | cut -d ' ' -f 1)
+# only first 14 columns of annotations table are consistent between runs
```
Capital ```O``` on only.

LeeBergstrand commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
+REFERENCE_FILE_HASHES[2]=$(cut -f 1-14 rpoB_0_MetagenomeTest_0_annotations.tsv | md5sum | cut -d ' ' -f 1)
 
 # Store generated test output hashes in array
+# Keys in REFERENCE_FILE_HASHES should correspond to same key in REGENERATED_FILE_HASHES
```
By ```keys``` do you mean ```elements```?

LeeBergstrand replied:
Also its good practice to not use variable names (unless you have to) in comments. The comments should use plain languages. "The reference file hashes should match the input file hashes."

LeeBergstrand replied:
The comments should explain what the code is trying to achieve. Not what it does.

LeeBergstrand replied:
Might want to remove this comment (just saw the below). The comment should be something like "Store the output file hashes for later comparison"

LeeBergstrand commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
-    if [[ ${entry} == *"refseq"* && ${entry} == *".fa"* ]]; then
-        cd test_output
+    # Checks if entry is the refseq msa fasta
+    if [[ ${entry} == rpoB_0_refseq_msa*fa ]]; then
```
The next two if's should be ```else if``` (https://ryanstutorials.net/bash-scripting-tutorial/bash-if-statements.php.). 
We know that if the file enters is ```rpoB_0_msa*fa``` then it can't be ```rpoB_0_refseq_msa*fa``` or ```rpoB_0_MetagenomeTest_0_annotations*tsv``` so we can not do those if statements and go to the next iteration of the loop.

LeeBergstrand commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
-    exit 1
-fi
+# If the hashes match, pass the test
+# Keys in REFERENCE_FILE_HASHES should correspond to same key in REGENERATED_FILE_HASHES
```
See comment above.

LeeBergstrand commented:
@kemendle https://circleci.com/gh/Metannotate/Metannotate/360 

Just noticed this on Circle CI.

```
Job ran successfully. The following files are now available:

/home/circleci/repo/test_output/rpoB_0_MetagenomeTest_0_reads_mu8w9O804247682.fa
/home/circleci/repo/test_output/rpoB_0_msa_engzbS31409321.fa
/home/circleci/repo/test_output/rpoB_0_refseq_msa_6gg91R875621562.fa
/home/circleci/repo/test_output/rpoB_0_refseq_tree_96MHjH498427594.newick
/home/circleci/repo/test_output/rpoB_0_refseq_fastree_log_oHoP3k598066503.txt
/home/circleci/repo/test_output/rpoB_0_tree_SDEGk155227671.newick
/home/circleci/repo/test_output/rpoB_0_MetagenomeTest_0_annotations_aZZZuf136245916.tsv
/home/circleci/repo/test_output/krona_xB0K1e509874450.html
/home/circleci/repo/test_output/raw_counts_xoGAq5126276164.csv
/home/circleci/repo/test_output/normalized_counts_LF37AN310807051.csv
Verifying outputs...
Correct number of files in output_dir (13)
rpoB_0_msa_engzbS31409321.fa passed the hash check.
rpoB_0_refseq_msa_6gg91R875621562.fa passed the hash check.
```

Was it supposed to run the hash check on the third file (```rpoB_0_MetagenomeTest_0_annotations*tsv```) as well?

jmtsuji commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
+        EXIT_CODE=1
+    fi
+done
+
```
Consider adding a final print statement like "Test passed" before finishing if `${EXIT_CODE} == 0`.

LeeBergstrand replied:
Has been addressed.

jmtsuji commented on testing/test_metannotate_end_to_end.sh:
```diff
@@ -14,38 +14,64 @@ python run_metannotate.py --orf_files=data/MetagenomeTest.fa --hmm_files=data/hm
+cd ${ROOT_PATH}/test_output
+for entry in $(ls); do
+    # Checks if entry is the query msa fasta
+    if [[ ${entry} == rpoB_0_msa*fa ]]; then
```
Note that the wildcard here (rpoB_0_msa*fa) could cause problems. I noticed when I tried to re-run the test script after the pipeline failed that multiple versions of each output file (each with a different hash where the wildcard is) were produced.

I propose checking that the test_output directory is empty before starting the automated test. This should solve this problem.

kmendler replied:
Should I have it start by removing any existing output folders with the same name (ie. test_output)? Alternatively, I can add a timestamp to the '--output_dir' folder name.

jmtsuji replied:
Maybe have the script error out if `test_output` already exists before the test starts. The error message could instruct the user to delete the folder before proceeding. Does this seem reasonable?

jmtsuji reviewed (changes requested):
Good overall. See the comment about checking that test_outputs is empty before starting the test.

Commit:
Modify test implementation and refactor

Add test for number of files in output folder
Add check for annotations table in output folder
Add test summary to indicate whether at least one test failed

Commit:
Check that output directory does not exist or is empty before running tests

LeeBergstrand reviewed (approved):
Looks good. Much improved. 💃
